connector/src/main/java/org/jboss/as/connector/subsystems/resourceadapters/ResourceAdapterParser.java

```diff
@@ -596,6 +600,121 @@ protected WorkManagerSecurity parseWorkManagerSecurity_5_0(final ModelNode opera
+                        }
+                        case USERS: {
+                            userMappingEnabled = true;
+                            break;
+                        }
+                        case GROUPS: {
+                            userMappingEnabled = false;
+                            break;
+                        }
+                        case MAP: {
+                            if (userMappingEnabled) {
+                                String from = rawAttributeText(reader, WorkManagerSecurity.Attribute.FROM.getLocalName());
+                                if (from == null || from.trim().equals(""))
+                                    throw new ParserException(bundle.requiredAttributeMissing(WorkManagerSecurity.Attribute.FROM.getLocalName(), reader.getLocalName()));
+                                String to = rawAttributeText(reader, WorkManagerSecurity.Attribute.TO.getLocalName());
+                                if (to == null || to.trim().equals(""))
+                                    throw new ParserException(bundle.requiredAttributeMissing(WorkManagerSecurity.Attribute.TO.getLocalName(), reader.getLocalName()));
+                                ModelNode object = new ModelNode();
+                                WM_SECURITY_MAPPING_FROM.parseAndSetParameter(from, object, reader);
+                                WM_SECURITY_MAPPING_TO.parseAndSetParameter(to, object, reader);
+                                operation.get(WM_SECURITY_MAPPING_USERS.getName()).add(object);
+                            } else {
+                                String from = rawAttributeText(reader, WorkManagerSecurity.Attribute.FROM.getLocalName());
+                                if (from == null || from.trim().equals(""))
+                                    throw new ParserException(bundle.requiredAttributeMissing(WorkManagerSecurity.Attribute.FROM.getLocalName(), reader.getLocalName()));
+                                String to = rawAttributeText(reader, WorkManagerSecurity.Attribute.TO.getLocalName());
+                                if (to == null || to.trim().equals(""))
+                                    throw new ParserException(bundle.requiredAttributeMissing(WorkManagerSecurity.Attribute.TO.getLocalName(), reader.getLocalName()));
+                                ModelNode object = new ModelNode();
+                                WM_SECURITY_MAPPING_FROM.parseAndSetParameter(from, object, reader);
+                                WM_SECURITY_MAPPING_TO.parseAndSetParameter(to, object, reader);
+                                operation.get(WM_SECURITY_MAPPING_GROUPS.getName()).add(object);
+                            }
+                            break;
+                        }
+                        default:
+                            throw new ParserException(bundle.unexpectedElement(reader.getLocalName()));
+                    }
+                    break;
+                }
+            }
```

Comment:
Why duplicate all of this logic?

Reply:
I believe I did this for the same reason why `parseWorkManagerSecurity_5_0` duplicates the `parseWorkManagerSecurity` (and adds the `ELYTRON_SECURITY_DOMAIN`), to be more flexible when it comes to future changes in the 7_0 parse method. Would you prefer if I created separate method for the duplicate parsing code and used it in the 5_0 and 7_0 parse methods??

Reply:
The reason early on we started using the duplication of parse methods for newer methods was to make it easier to just delete old versions as they became unsupported rather than having to manually unravel version specific changes within the method.

Reply:
In general, inheritance is a poor way to model versioning...
But given the current state, it would make sense to encapsulate the parsing of the attributes of the security element in a separate method from the parsing of the child elements.  This way your subclass only needs to override the attribute parsing, and eliminate the duplication of the child element parsing.